File: scratch/compact-guards-integration-20260916/compact/src/Test/PolymorphismFFI.erl
-module(test_polymorphismFFI@foreign).
-export([runPolymorphismFFI/1]).
runPolymorphismFFI(N) ->
  Dict = #{mempty => 1, mappend => fun(X) -> fun(Y) -> X + Y end end},
  poly_loop(N, Dict, 0).
poly_loop(0, _Dict, Acc) -> Acc;
poly_loop(N, Dict, Acc) ->
  Append = maps:get(mappend, Dict),
  poly_loop(N - 1, Dict, (Append(Acc))(maps:get(mempty, Dict))).

File: src/Test/PolymorphismFFI.java
    private record Monoidish<A>(A mempty, java.util.function.Function<A, java.util.function.Function<A, A>> mappend) {}

    private static <A> A polyLoop(Monoidish<A> dictionary, int count, A initial) {
        A result = initial;
        while (count > 0) {
            result = dictionary.mappend().apply(result).apply(dictionary.mempty());
            count--;
        }
        return result;
    }

    public static final java.util.function.Function<Object, Object> runPolymorphismFFI = Bench.nativeBenchmark(input ->
        polyLoop(new Monoidish<>(1, left -> right -> left + right), (Integer) input, 0));
